### PR TITLE
operator: Add cilium node garbage collector

### PR DESCRIPTION
**Pull request by sayboras**
### Description

In the normal scenario, CiliumNode is created by agent with owner
references attached all time time in below PR[0]. However, there could
be the case that CiliumNode is created by IPAM module[1], which
didn't have any ownerReferences at all. For this case, if the
corresponding node got terminated and never came back with same
name, the CiliumNode resource is still dangling, and needs to be
garbage collected.

This commit is to add garbage collector for CiliumNode, with below
logic:

- Gargage collector will run every predefined interval (e.g. specify
  by flag --nodes-gc-interval)
- Each run will check if CiliumNode is having a counterpart k8s node
  resource. Also, remove this node from GC candidate if required.
- If yes, CiliumNode is considered as valid, happy day.
- If no, check if ownerReferences are set.
  - If yes, let k8s perform garbage collection.
  - If no, mark the node as GC candidate. If in the next run, this
    node is still in GC candidate, remove it.

### Testing

Testing was done locally with kind cluster

<details>
<summary>Create one dummy ciliumnode and check gc </summary>

```
$ ksyslo deployment/cilium-operator --timestamps | grep kind-worker-dummy
2022-04-26T10:59:22.106774489Z level=debug msg="UpdatedStatus Node" error="<nil>" node-name=kind-worker-dummy subsys=pod-cidr
2022-04-26T10:59:23.087007602Z level=debug msg="UpdatedStatus Node" error="<nil>" node-name=kind-worker-dummy subsys=pod-cidr
2022-04-26T10:59:51.977272944Z level=info msg="Add CiliumNode to garbage collector candidates" nodeName=kind-worker-dummy subsys=watchers
2022-04-26T11:00:21.978387751Z level=info msg="Perform GC for invalid CiliumNode" nodeName=kind-worker-dummy subsys=watchers

$ kg ciliumnodes
NAME                 AGE
kind-control-plane   156m
kind-worker          156m
```

</details>

### CODE REVIEW

File: install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml [hunk 1]
```diff
@@ -25,8 +25,8 @@ rules:
 - apiGroups:
   - ""
   resources:
-{{- if .Values.operator.removeNodeTaints }}
-  # To remove node taints
+{{- if or .Values.operator.removeNodeTaints .Values.operator.nodeGCInterval }}
+  # To remove node taints or perform underlying CiliumNode garbage collector
```
Comment (christarazi) [inline install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml]:
```suggestion
  # To remove node taints or perform underlying CiliumNode garbage collection
```

File: Documentation/operations/upgrade.rst [hunk 2]
```diff
@@ -326,6 +326,9 @@ New Options
   for EC2 instances (and existing ENIs). EC2 Instances tags is in the form of k1=v1,k2=v2
   (multiple k/v pairs can also be passed by repeating the CLI flag")
   Use ``instanceTagsFilter`` in Helm chart.
+* ``nodes-gc-interval``: This option was marked as deprecated and has no effect
+  in 1.11. Cilium Node Garbage collector is added back in 1.12, so this flag
+  can be moved out of deprecated list.
```
Comment (christarazi) [inline Documentation/operations/upgrade.rst]: This option used to exist for kvstore GC. Now we are using it to implement K8s GC, so we should point that out.

File: operator/flags.go [hunk 3]
```diff
@@ -249,10 +249,8 @@ func init() {
 	flags.String(option.K8sKubeConfigPath, "", "Absolute path of the kubernetes kubeconfig file")
 	option.BindEnv(option.K8sKubeConfigPath)
 
-	// To be removed in Cilium 1.12
-	flags.Duration(operatorOption.NodesGCInterval, 2*time.Minute, "GC interval for nodes store in the kvstore")
+	flags.Duration(operatorOption.NodesGCInterval, 5*time.Minute, "GC interval for nodes store in the kvstore")
```
Comment (christarazi) [inline operator/flags.go]: We are not doing kvstore GC anymore.

File: operator/watchers/cilium_node_gc.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+		}
+
+		if !k8serrors.IsNotFound(err) {
+			scopedLog.WithError(err).Error("Unable to fetch k8s node from store")
+			return err
+		}
+
+		obj, _, err := ciliumNodeStore.GetByKey(nodeName)
+		if err != nil {
+			scopedLog.WithError(err).Error("Unable to fetch cilium node from store")
```
Comment (christarazi) [inline operator/watchers/cilium_node_gc.go]: Nit: there's a mix of usages between "ciilum node" and "CiliumNode". I suggest we stick with the latter everywhere.

**Review by christarazi — CHANGES_REQUESTED** [review]
Changes look good, just a few minor comments

File: operator/flags.go (hunk 3 above)
Comment (sayboras) [reply operator/flags.go]: thanks :100:

File: operator/watchers/cilium_node_gc.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+func (c *ciliumNodeGCCandidate) Delete(nodeName string) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	delete(c.nodesToRemove, nodeName)
+}
+
+// RunCiliumNodeGC performs garbage collector for cilium node resource
+func RunCiliumNodeGC(ctx context.Context, ciliumNodeStore cache.Store, interval time.Duration, stopCh <-chan struct{}) {
+	nodesInit(k8s.WatcherClient(), stopCh)
```
Comment (kaworu) [inline operator/watchers/cilium_node_gc.go]: Admittedly I'm lacking context about how `nodesInit` uses `stopCh`, but the caller never close it and is seems like a strange pattern to me. Would it make sense to pass `ctx.Done()` here instead?

**Review by kaworu — CHANGES_REQUESTED** [review]
Seeing that we are selectively adding RBAC permissions and passing the `nodes-gc-interval` flag through the ConfigMap, I'm assuming there is a use-case for disabling this feature. In what context users would want to disable it?

Follow-up question: how are users expected to effectively disable this feature through Helm? I think the combination of the flag being a duration, the default value, and different ways of checking it (`hasKey` in the `ConfigMap`, "emptiness" in the `ClusterRole`) might not work as expected:

```console
% helm template cilium ./install/kubernetes/cilium --namespace kube-system | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: "5m0s"
    # To perform CiliumNode garbage collector
% helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval=1s | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: "1s"
    # To perform CiliumNode garbage collector
```

Here the first `grep` expression is for the flag value, second expression for RBAC. Using the default and override will correctly set the flag and add the RBAC permission, happy day™.

```console
% helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval=null | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
% helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval= | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: ""
% helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval=0s | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: "0s"
    # To perform CiliumNode garbage collector
```

Settting to `null` shows nothing, meaning we don't pass the flag and we don't have the RBAC permission. The issue here is that we'll be using the non-zero CLI default (5m) and will fail because we don't have the permission.

Settting to empty will actually pass it as-is to the flag, which will fail duration parsing.

Settting to `0s` will correctly disable the feature in the ConfigMap, but won't add the RBAC permission.

**Review by christarazi — APPROVED** [review]

**Comment by sayboras** [comment]
Thanks for your input, indeed there are a few cases I didn't consider.

> Seeing that we are selectively adding RBAC permissions and passing the nodes-gc-interval flag through the ConfigMap, I'm assuming there is a use-case for disabling this feature. In what context users would want to disable it?

Normally, we don't want to disable this GC, it's just an option for users in case of any issue later on.

> Settting to null shows nothing, meaning we don't pass the flag and we don't have the RBAC permission. The issue here is that we'll be using the non-zero CLI default (5m) and will fail because we don't have the permission.
Settting to empty will actually pass it as-is to the flag, which will fail duration parsing.
Settting to 0s will correctly disable the feature in the ConfigMap, but won't add the RBAC permission.

Agreed on the above points, I have made below changes to rectify
- update default value in CLI to 0s i.e disabled by default, so that we don't couple with RBAC permission
- add default value in config map
- add check in RBAC to make sure that we only add permission if .Values.operator.nodeGCInterval is having some value, and it's not 0s. I don't think we need to check for correct data type (e.g. duration) in helm.

<details>

```
$ helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval=null | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"

$ helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval= | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector" 
  nodes-gc-interval: "0s"
 
$ helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval=0s | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: "0s"
  
$ helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval=1s | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: "1s"
    # To perform CiliumNode garbage collector
  
```

</details>

**Review by kaworu — APPROVED** [review]
Thanks @sayboras, patch LGTM now although IMO it would be a tad better if the helm default reflect the cli default. Nitpicking, but here is a suggestion:
<details>
  <summary>details</summary>

```diff
diff --git a/install/kubernetes/cilium/templates/cilium-configmap.yaml b/install/kubernetes/cilium/templates/cilium-configmap.yaml
index 9dabbc381e..fc03209faf 100644
--- a/install/kubernetes/cilium/templates/cilium-configmap.yaml
+++ b/install/kubernetes/cilium/templates/cilium-configmap.yaml
@@ -109,8 +109,10 @@ data:
   cilium-endpoint-gc-interval: "{{ .Values.operator.endpointGCInterval }}"
 {{- end }}
 
-{{- if hasKey .Values.operator "nodeGCInterval" }}
-  nodes-gc-interval: "{{ .Values.operator.nodeGCInterval | default "0s" }}"
+{{- if .Values.operator.nodeGCInterval }}
+  nodes-gc-interval: {{ .Values.operator.nodeGCInterval | quote }}
+{{- else }}
+  nodes-gc-interval: 0
 {{- end }}
 
 {{- if hasKey .Values "disableEndpointCRD" }}
diff --git a/install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml b/install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml
index 8af8b9824b..78e8d27c9a 100644
--- a/install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml
+++ b/install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml
@@ -1,4 +1,5 @@
 {{- if and .Values.operator.enabled .Values.serviceAccounts.operator.create }}
+{{- $nodeGC := and .Values.operator.nodeGCInterval (not (regexMatch "^(?:0(?:ns|[uµ]s|ms|s|m|h)?)+$" .Values.operator.nodeGCInterval)) }}
 apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRole
 metadata:
@@ -25,7 +26,7 @@ rules:
 - apiGroups:
   - ""
   resources:
-{{- if or .Values.operator.removeNodeTaints .Values.operator.nodeGCInterval }}
+{{- if or .Values.operator.removeNodeTaints $nodeGC }}
   # To remove node taints or perform underlying CiliumNode garbage collection
   - nodes
 {{- end }}
@@ -123,12 +124,10 @@ rules:
   - get
   - list
   - watch
-{{- if .Values.operator.nodeGCInterval }}
-{{- if ne .Values.operator.nodeGCInterval "0s" }}
+{{- if $nodeGC }}
     # To perform CiliumNode garbage collector
   - delete
 {{- end }}
-{{- end }}
 - apiGroups:
   - cilium.io
   resources:
diff --git a/operator/flags.go b/operator/flags.go
index d1f11f2aa0..a3e79a2e6d 100644
--- a/operator/flags.go
+++ b/operator/flags.go
@@ -249,7 +249,7 @@ func init() {
 	flags.String(option.K8sKubeConfigPath, "", "Absolute path of the kubernetes kubeconfig file")
 	option.BindEnv(option.K8sKubeConfigPath)
 
-	flags.Duration(operatorOption.NodesGCInterval, 0*time.Second, "GC interval for CiliumNodes")
+	flags.Duration(operatorOption.NodesGCInterval, 5*time.Minute, "GC interval for CiliumNodes")
 	option.BindEnv(operatorOption.NodesGCInterval)
 
 	flags.String(operatorOption.OperatorPrometheusServeAddr, operatorOption.PrometheusServeAddr, "Address to serve Prometheus metrics")
```
</details>

With that patch we have the cli default to 5m.

`null`, empty, and `0` yield the same template:

```console
% helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval=null | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: 0
% helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval= | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: 0
% helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval=0 | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: 0
```

We attempt to detect a user-provided zero value, which we pass on the `ConfigMap` so that the user find may back their provided value (although they have no effect):

```console
% helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval=0ns | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: "0ns"
% helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval=0us | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: "0us"
% helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval=0µs | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: "0µs"
% helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval=0ms | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: "0ms"
% helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval=0s | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: "0s"
% helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval=0m | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: "0m"
% helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval=0h | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: "0h"
% helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval=0h0m00s | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: "0h0m00s"
```

Of course override and default should be passed to the `ConfigMap` and enable RBAC:

```console
% helm template cilium ./install/kubernetes/cilium --namespace kube-system --set operator.nodeGCInterval=42s | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: "42s"
    # To perform CiliumNode garbage collector
% helm template cilium ./install/kubernetes/cilium --namespace kube-system | grep -E "nodes-gc-interval|# To perform CiliumNode garbage collector"
  nodes-gc-interval: "5m0s"
    # To perform CiliumNode garbage collector
```

**Review by qmonnet — APPROVED** [review]

**Comment by sayboras** [comment]
/test-1.23-net-next

**Comment by sayboras** [comment]
Thanks Alex.

I noticed that this suggestion should be applied for other duration field in helm values as well. For consistency, I think we can just keep as what we have in this PR.

Raised one https://github.com/cilium/cilium/issues/19684 for future improvement.

**Comment by sayboras** [comment]
All tests passed except NAT-46x64 and ci-l4lb jobs are still under waiting, there two jobs are going to merge to one, and they are not related to this change.

Mark PR ready to merge.

**Comment by aanm** [comment]
waiting on https://github.com/cilium/cilium/pull/19686 to be merged

**Comment by sayboras** [comment]
Force push to resolve the conflict with master, only one file is changed, the diff can be found as per below:

```diff
diff --git a/install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml b/install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml
index 531566cc4b..8af8b9824b 100644
--- a/install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml
+++ b/install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml
@@ -21,21 +21,12 @@ rules:
 {{- end }}
 {{- end }}
 {{- end }}
-{{- if or .Values.operator.removeNodeTaints .Values.operator.setNodeNetworkStatus .Values.operator.endpointGCInterval }}
-- apiGroups:
-  - ""
-  resources:
-  - nodes
-  verbs:
-  - list
-  - watch
-{{- end }}
 {{- if or .Values.operator.removeNodeTaints .Values.operator.setNodeNetworkStatus }}
 - apiGroups:
   - ""
   resources:
-{{- if .Values.operator.removeNodeTaints }}
-  # To remove node taints
+{{- if or .Values.operator.removeNodeTaints .Values.operator.nodeGCInterval }}
+  # To remove node taints or perform underlying CiliumNode garbage collection
   - nodes
 {{- end }}
 {{- if .Values.operator.setNodeNetworkStatus }}
@@ -43,6 +34,8 @@ rules:
   - nodes/status
 {{- end }}
   verbs:
+  - list
+  - watch
   - patch
 {{- end }}
 - apiGroups:
```

**Comment by sayboras** [comment]
/test-1.16-4.9

**Comment by sayboras** [comment]
/test-1.22-4.19

**Comment by sayboras** [comment]
/test-1.23-net-next

**Comment by sayboras** [comment]
/test-1.21-5.4

**Comment by sayboras** [comment]
All the tests passed now, marking this ready for merge.

**Pull request merged**